### PR TITLE
Join point clouds

### PR DESCRIPTION
**Pull request by pmokeev**

### CODE REVIEW

File: src/algorithmForNN/pointCloudUtils.py [hunk 1]
```diff
@@ -151,6 +151,104 @@ def get_area_of_plane(points: o3d.geometry.PointCloud, plane_model: list) -> flo
...
+            calib_matrix[2, 1],
+            calib_matrix[2, 2],
+            calib_matrix[2, 3],
+        ) = list(map(float, file.readlines()[4][4:].rstrip().split(" ")))
+        calib_matrix[3, 3] = 1
+
+    return calib_matrix
+
+
+def get_position_from_poses_file(
```
Comment (DmiitriyJarosh) [inline src/algorithmForNN/pointCloudUtils.py:176]: matrix word is skipped

File: src/algorithmForNN/pointCloudUtils.py [hunk 2]
```diff
@@ -151,6 +151,104 @@ def get_area_of_plane(points: o3d.geometry.PointCloud, plane_model: list) -> flo
...
+            pose_matrix[2, 2],
+            pose_matrix[2, 3],
+        ) = list(map(np.float64, lines[frame_number].rstrip().split(" ")))
+        pose_matrix[3, 3] = 1
+
+    return pose_matrix
+
+
+def transform_positions_in_point_cloud(
+    calib_matrix: np.ndarray, pose_matrix: np.ndarray, point_claud: np.ndarray
```
Comment (DmiitriyJarosh) [inline src/algorithmForNN/pointCloudUtils.py:202]: point_clOud

File: src/algorithmForNN/pointCloudUtils.py [hunk 3]
```diff
@@ -151,6 +151,104 @@ def get_area_of_plane(points: o3d.geometry.PointCloud, plane_model: list) -> flo
...
+) -> np.ndarray:
+    left_camera_matrix = np.matmul(pose_matrix, calib_matrix)
+
+    for index, point in enumerate(point_claud):
+        temp_point = np.array([[point[0], point[1], point[2], 1]]).T
+        new_point = np.matmul(left_camera_matrix, temp_point)
+        point_claud[index, 0], point_claud[index, 1], point_claud[index, 2] = (
```
Comment (DmiitriyJarosh) [inline src/algorithmForNN/pointCloudUtils.py:209]: can this transform be done with o3d built-in function for pcd transform?

File: src/algorithmForNN/pointCloudUtils.py [hunk 4]
```diff
@@ -151,6 +151,104 @@ def get_area_of_plane(points: o3d.geometry.PointCloud, plane_model: list) -> flo
...
+            new_point[0, 0],
+            new_point[1, 0],
+            new_point[2, 0],
+        )
+
+    return point_claud
+
+
+def create_point_cloud_by_first_N_snapshots(
```
Comment (DmiitriyJarosh) [inline src/algorithmForNN/pointCloudUtils.py:218]: i think it is not an util function

File: src/algorithmForNN/pointCloudUtils.py [hunk 5]
```diff
@@ -151,6 +151,104 @@ def get_area_of_plane(points: o3d.geometry.PointCloud, plane_model: list) -> flo
...
+                os.path.join(
+                    path_to_bin_dir, "0" * (6 - len(str(index))) + str(index) + ".bin"
+                )
+            )
+        )
+        temp_point_cloud = transform_positions_in_point_cloud(
+            calib_matrix, pose_matrix, temp_point_cloud
+        )
+
+        main_point_cloud = np.concatenate((main_point_cloud, temp_point_cloud))
```
Comment (DmiitriyJarosh) [inline src/algorithmForNN/pointCloudUtils.py:247]: lets move all to loop - as we can start with empty cloud before it

**Review by DmiitriyJarosh — CHANGES_REQUESTED** [review]

File: src/algorithmForNN/pointCloudUtils.py (hunk 4 above)
Comment (pmokeev) [reply src/algorithmForNN/pointCloudUtils.py:218]: where do you suggest to move this function?

File: src/algorithmForNN/pointCloudUtils.py [hunk 6]
```diff
@@ -151,6 +151,111 @@ def get_area_of_plane(points: o3d.geometry.PointCloud, plane_model: list) -> flo
     return current_list[1] * current_list[2]
 
 
+def get_calibration_matrix_from_calib_file(path_to_calibration_file: str) -> np.ndarray:
+    calib_matrix = np.zeros((4, 4))
+    with open(path_to_calibration_file) as file:
+        (
```
Comment (anastasiia-kornilova) [inline src/algorithmForNN/pointCloudUtils.py:157]: Oh, looks crazy... We don't pay for number of lines... :)))

Firstly, you can initialize drectly with `calib_matrix = np.eye(4)`

Secondly

`calib_matrix[:3,:4] = np.array(list(map(float, file.readlines()[4][4:].rstrip().split(" ")))).reshape(3, 4)` , check, maybe it is fine even without call of "list" inside.

File: src/algorithmForNN/pointCloudUtils.py [hunk 7]
```diff
@@ -151,6 +151,111 @@ def get_area_of_plane(points: o3d.geometry.PointCloud, plane_model: list) -> flo
...
+        ) = list(map(float, file.readlines()[4][4:].rstrip().split(" ")))
+        calib_matrix[3, 3] = 1
+
+    return calib_matrix
+
+
+def get_position_matrix_from_poses_file(
+    path_to_poses_file: str, frame_number: int
+) -> np.ndarray:
+    pose_matrix = np.zeros((4, 4))
```
Comment (anastasiia-kornilova) [inline src/algorithmForNN/pointCloudUtils.py:179]: Same?

Comment (anastasiia-kornilova) [reply src/algorithmForNN/pointCloudUtils.py:179]: Better to read poses all together one time opening file, not to access storage every time. They are quite light.

File: src/algorithmForNN/pointCloudUtils.py [hunk 8]
```diff
@@ -151,6 +151,111 @@ def get_area_of_plane(points: o3d.geometry.PointCloud, plane_model: list) -> flo
...
+
+    return pose_matrix
+
+
+def transform_positions_in_point_cloud(
+    calib_matrix: np.ndarray,
+    pose_matrix: np.ndarray,
+    point_cloud: o3d.geometry.PointCloud,
+) -> o3d.geometry.PointCloud:
+    left_camera_matrix = np.matmul(pose_matrix, calib_matrix)
```
Comment (anastasiia-kornilova) [inline src/algorithmForNN/pointCloudUtils.py:206]: `left_camera_matrix = pose_matrix @ calib_matrix`

File: src/algorithmForNN/pointCloudUtils.py [hunk 9]
```diff
@@ -151,6 +151,111 @@ def get_area_of_plane(points: o3d.geometry.PointCloud, plane_model: list) -> flo
...
+
+    return point_cloud.transform(left_camera_matrix)
+
+
+def merge_two_point_clouds(
```
Comment (anastasiia-kornilova) [inline src/algorithmForNN/pointCloudUtils.py:211]: open3d has `+` operator for point clouds

File: src/algorithmForNN/pointCloudUtils.py [hunk 10]
```diff
@@ -151,6 +151,111 @@ def get_area_of_plane(points: o3d.geometry.PointCloud, plane_model: list) -> flo
...
+    path_to_bin_dir = os.path.join(path_to_dataset, "velodyne")
+    path_to_poses_file = os.path.join(path_to_dataset, "poses.txt")
+    path_to_calib_file = os.path.join(path_to_dataset, "calib.txt")
+
+    point_cloud = o3d.geometry.PointCloud()
+    calib_matrix = get_calibration_matrix_from_calib_file(path_to_calib_file)
+
+    for index in range(first_number_of_point_cloud, last_number_of_point_cloud + 1):
+        path_to_current_point_cloud = os.path.join(
+            path_to_bin_dir, "0" * (6 - len(str(index))) + str(index) + ".bin"
```
Comment (anastasiia-kornilova) [inline src/algorithmForNN/pointCloudUtils.py:246]: `zfill` instead of ugly zeros copying?

**Review by anastasiia-kornilova — CHANGES_REQUESTED** [review]

File: src/algorithmForNN/pointCloudUtils.py (hunk 4 above)
Comment (DmiitriyJarosh) [reply src/algorithmForNN/pointCloudUtils.py:232]: Some special file --- where algo of pcd combining will be

File: src/algorithmForNN/pointCloudUtils.py [hunk 11]
```diff
@@ -151,6 +151,48 @@ def get_area_of_plane(points: o3d.geometry.PointCloud, plane_model: list) -> flo
     return current_list[1] * current_list[2]
 
 
+def get_calibration_matrix_from_calib_file(path_to_calibration_file: str) -> np.ndarray:
+    calib_matrix = np.eye(4)
+    with open(path_to_calibration_file) as file:
```
Comment (anastasiia-kornilova) [inline src/algorithmForNN/pointCloudUtils.py:156]: Looks like this code reused twice, let's create io function related to kitti dataset that constructs 4x4 transformation matrix from kitti lines.

File: src/algorithmForNN/combinePointClouds.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    for index in range(first_number_of_point_cloud, last_number_of_point_cloud + 1):
+        path_to_current_point_cloud = os.path.join(
+            path_to_bin_dir, str(index).zfill(6) + ".bin"
+        )
+        temp_point_cloud = get_point_cloud_from_bin_file(path_to_current_point_cloud)
+        temp_point_cloud = transform_positions_in_point_cloud(
+            calib_matrix, poses_matrices[index], temp_point_cloud
+        )
+
+        point_cloud = merge_two_point_clouds(point_cloud, temp_point_cloud)
```
Comment (anastasiia-kornilova) [inline src/algorithmForNN/combinePointClouds.py:31]: Why do you need special function with huge name that overwrites cute `+`?

**Review by anastasiia-kornilova — CHANGES_REQUESTED** [review]

**Review by anastasiia-kornilova — APPROVED** [review]

**Pull request merged**